### PR TITLE
Vulkan: Implement stencil upload (for Star Ocean).

### PR DESCRIPTION
**Pull request by hrydgard**
Does it the good old way. Might investigate potentially faster methods at a later time.

### CODE REVIEW

**Comment by DJScias** [comment]
Hello there,

Just a curiosity, but is the 'stencil' stuff the reason why characters were invisible in Vulkan? I had to change to D3D11 and use buffered rendering (as Vulkan doesn't support this yet) before they would show up.

Or is this another issue and I am confusing them?

Kind regards,
DJScias

**Comment by hrydgard** [comment]
In the case of Star Ocean, yes. It works now.

**Comment by DJScias** [comment]
I'm currently playing through these games (1&2) and can let you know with an issue if it still persists.

Just to check: does Vulkan work with buffered rendering now or should it work without that now as well on Vulkan?

Kind regards,
DJScias

**Comment by hrydgard** [comment]
Buffered rendering works now, I'd recommend keeping it on.

**Comment by DJScias** [comment]
Thanks for the information and good work on the emulator, it's been amazing so far.

Currently compiling my own windows build to check as the build bot seems to have not done anything since the 29th of October.

**Comment by hrydgard** [comment]
Yeah, I'll look into what's up with the buildbot..

File: GPU/Vulkan/StencilBufferVulkan.cpp [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+#include "ext/native/thin3d/VulkanRenderManager.h"
+#include "Core/Reporting.h"
+#include "GPU/Vulkan/FramebufferVulkan.h"
+#include "GPU/Vulkan/FragmentShaderGeneratorVulkan.h"
+#include "GPU/Vulkan/ShaderManagerVulkan.h"
+#include "GPU/Vulkan/TextureCacheVulkan.h"
+#include "GPU/Vulkan/VulkanUtil.h"
+
+#define STR_HELPER(x) #x
+#define STR(x) STR_HELPER(x)
```
Comment (unknownbrackets) [inline GPU/Vulkan/StencilBufferVulkan.cpp]: nit: Not used.  DX9 thing from register ids.

-[Unknown]

File: GPU/Vulkan/StencilBufferVulkan.cpp [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+};
+layout (location = 0) in vec2 v_texcoord0;
+layout (location = 0) out vec4 fragColor0;
+
+void main() {
+  vec4 index = texture(tex, v_texcoord0);
+	int indexBits = int(floor(index.a * 255.99)) & 0xFF;
+	if ((indexBits & u_stencilValue) == 0)
+		discard;
+  fragColor0 = index.aaaa;
```
Comment (unknownbrackets) [inline GPU/Vulkan/StencilBufferVulkan.cpp]: nit: Tabs and spaces living in intolerable harmony (also stencil_vs.)

-[Unknown]

File: GPU/Vulkan/StencilBufferVulkan.cpp [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+
+	VulkanRenderManager *renderManager = (VulkanRenderManager *)draw_->GetNativeObject(Draw::NativeObject::RENDER_MANAGER);
+
+	if (usedBits == 0) {
+		if (skipZero) {
+			// Common when creating buffers, it's already 0.  We're done.
+			return false;
+		}
+
+		// TODO: Find a nice way to clear alpha here too.
```
Comment (unknownbrackets) [inline GPU/Vulkan/StencilBufferVulkan.cpp]: VulkanRenderManager::Clear?  I know it's ideally avoided, but, here we want to just clear one channel so...

-[Unknown]

File: GPU/Vulkan/VulkanUtil.cpp [hunk 4]
```diff
@@ -208,7 +210,7 @@ VkPipeline Vulkan2D::GetPipeline(VkRenderPass rp, VkShaderModule vs, VkShaderMod
 
 	VkPipelineColorBlendAttachmentState blend0 = {};
 	blend0.blendEnable = false;
-	blend0.colorWriteMask = 0xF;
+	blend0.colorWriteMask = depthStencilMode == VK2DDepthStencilMode::STENCIL_REPLACE_ALWAYS ? 0 : 0xF;
```
Comment (unknownbrackets) [inline GPU/Vulkan/VulkanUtil.cpp]: For stencil upload, we want `VK_COLOR_COMPONENT_A_BIT`, no?  That's what GLES does.

-[Unknown]

**Review by unknownbrackets — COMMENTED** [review]
Oops, GPU/Source.cpp.  I've done that before, heh...

-[Unknown]

File: GPU/Vulkan/StencilBufferVulkan.cpp (hunk 1 above)
Comment (hrydgard) [reply GPU/Vulkan/StencilBufferVulkan.cpp]: Yup. The uniform CB right after was also a leftover, killed that too.

File: GPU/Vulkan/StencilBufferVulkan.cpp (hunk 3 above)
Comment (hrydgard) [reply GPU/Vulkan/StencilBufferVulkan.cpp]: Clear can't clear just alpha. RGBA channel masks don't actually work on clears on a lot of hardware and are not present for clears in Vulkan, I assume drivers tend to emulate glClear with masks with a draw call. 

So yeah, I'm gonna have to do a draw call here.

File: GPU/Vulkan/VulkanUtil.cpp (hunk 4 above)
Comment (hrydgard) [reply GPU/Vulkan/VulkanUtil.cpp]: Oh, oops. Yes indeed.

File: GPU/Vulkan/StencilBufferVulkan.cpp (hunk 3 above)
Comment (unknownbrackets) [reply GPU/Vulkan/StencilBufferVulkan.cpp]: Ah right, yes.

-[Unknown]

**Pull request merged**